02-peer-protocol.md

```diff
@@ -390,6 +400,377 @@ would create a Denial of Service risk; therefore, forgetting it is recommended
+concatenated revocation points, and will not update. This allows stable
+channel tracking across RBFs and splicing attempts.
+
+### The `funding_signed2` Message
+
+This message exchanges the witness data for the inputs that were
+originally sent in the `funding_compose` message.
+
+1. type: 60 (`funding_signed2`)
+2. data:
+  [`32`:`channel_id`]
+  [`2`:`num_witnesses`]
+  [`num_witnesses`\*`witness_stack`]
+
+1. subtype: `witness_stack`
+2. data:
+  * [`2`:`num_input_witness`]
+  * [`num_input_witness`\*`witness_element`]
+
+1. subsubtype: `witness_element`
+2. data:
+  * [`2`:`len`]
+  * [`len`:`witness`]
+
+#### Requirements
+The sending node:
+  - MUST set `witness` to the marshalled witness data for each of its
+    inputs, in funding transaction order. FIXME: add reference
+  - MUST remember the details of this funding transaction.
+  - MUST NOT send a `witness_stack` whose length exceeds the
+    corresponding `max_witness_len`.
+  - if they have NOT received a valid `commitment_signed` message:
+    - MUST NOT send a `funding_signed2` message.
+
+The receiving node:
+  - if the `witness_stack` length exceeds the corresponding `max_witness_len`:
+    - MUST error.
+  - if the recipient was the sender of `open_channel2`:
+    - SHOULD reply with its own `funding_signed2`.
+  - SHOULD apply `witness`es to the funding transaction and
+    broadcast the result.
+
+#### Rationale
+Exchanging witness data allows both sides to broadcast the funding
+transaction and maintains information symmetry.
+
+### The `funding_locked2` Message
+
+1. type: 61 (`funding_locked2`)
+2. data:
+  * [`32`:`channel_id`]
+  * [`33`:`next_per_commitment_point`]
+
+#### Requirements and Rationale
+
+See [`funding_locked`](#the-funding_locked-message).
+
+
+### Kicking Off Replace-By-Fee: `init_rbf`
+
+This message is sent by the channel opener, after the funding
+transaction has been broadcast but before the `funding_locked2` message
+has been exchanged.
+
+Once an `init_rbf` message has been successfully ack'd by the accepter
+node, the message flow returns to `commitment_signed` and proceeds as
+indicated above.
+
+1. type: 62 (`init_rbf`)
+2. data:
+  * [`32`:`channel_id`]
+  * [`8`:`channel_reserve_satoshis`]
+  * [`4`:`feerate_per_kw`]
+  * [`4`:`feerate_per_kw_funding`]
+  * [`2`:`num_inputs`]
+  * [`num_inputs`\*`input_info`]
+  * [`2`:`num_outputs`]
+  * [`num_outputs`\*`output_info`]
+
+#### Requirements
+
+The sending node:
+  - MUST have sent `open_channel2`.
+  - MUST send a `feerate_per_kw_funding` greater than the previously
+    negotiated rate.
+  - MUST update the `channel_reserve_satoshis` to reflect the new input balance.
+  - MAY update the `feerate_per_kw`, the commitment transaction feerate.
+  - MAY include new, additional inputs.
+  - MAY set the `num_inputs` to zero.
+  - MUST transmit all outputs.
+  - MUST include at least one output with the `satoshis` set to zero,
+    to be used as the change output.
```

Comment:
Maybe review this depending on decision about non-change channels above...

Reply:
Updated to be 'optional' as well.